[WIP] Log marker

cyucelen opened the pull request:
It stands to reason that one of the strongest use cases would be for users to colorize their logs for easy reading.

- [x] Build a abstraction structure upon Golang's logger to colorize. (We can experiment here!)
- [x] Implement io.Writer
- [x] Add Mark Rules (_Mark Rule_ name can change)
    - [x] Builder way
    - [x] Passing Rule slice
- [x] README documentation

We can add further features to the checklist!

Closes #17

jnatalzia commented on log.go:
```diff
@@ -0,0 +1,48 @@
+
+func (l *LogMarker) Print(v ...interface{}) {
+	marked := fmt.Sprint(v...)
+	for _, rule := range l.rules {
```
How do you feel about the "constructor" for the `LogMarker` taking in a `MarkBuilder` instead? that way there's a single piece of the application that is always responsible for marking text, and you can avoid the logger having to maintain any state with regards to marking rules.

cyucelen replied:
`MarkBuilder` marks the text immediately and returns the marked text when finally `Build` called, so it doesn't store the mathcer functions or colors. I couldn't think the usage of `MarkBuilder` for applying same rules over log stream.

`Mark` function in `marker.go` file is the only responsible for marking text anyway.

Therefore, even if we use `MarkBuilder` for marking (we are just looping over rules and applying them instead of using `MarkBuilder`), we must store the marking rules specified by user to apply them for every log.

Lets also think about the UX of the log marker API, in my opinion we need to provide a lean interface with minimum amount of boilerplate code to the user. Can you give me an example usage with `MarkBuilder` from the user perspective("constructor" parameters etc.) for me to understand better, like (current implementation usage):

```golang
red := color.New(color.FgRed)
blue := color.New(color.FgBlue)
// function names most probably will change, take it as pseudo :P
logger := marker.NewLogger() 
logger.AddRule(marker.Rule{marker.MatchBracketSurrounded(), red}).
         AddRule(marker.Rule{marker.MatchParensSurrounded(), blue}) // just stores the rules
// Implementing AddRules([]marker.Rule) function can be beneficial to reduce struct literal boilerplate 
logger.Print("[Error] we have errors and we can see them!") // applies the stored rules before sending the log to the downstream log.Print function
logger.Print("(WOW) marker is [awesome] !") // applies to this too!
```

Correct me if I understood you wrong.

jnatalzia replied:
You understood me correctly! I had a misunderstanding about the way the builder operated. Makes perfect sense now :+1:. In suggesting the `MarkBuilder` I was trying to find a way where the actual logic for marking the strings lives in a central place and the logger is only responsible for facilitation. In thinking over it some more, the `Mark` and `MatchFunc` methods take care of this concern nicely so no need to over-engineer it 👍 

Approach makes sense, and thanks for the explanation!

cyucelen replied:
Cool! If we are agreed on the implementation way, we can start to implement and let the evolution make its job on the way!  :blush: 

Happy to see you as collaborator :tada: 

edit: also you can close the conversation for this review if we are agreed.

cyucelen commented:
Do we need to create a variadic `AddRule` method?
What do you think about its usefulness?

Darkclainer commented:
I think it should be another PR. See issue #27.

cyucelen commented:
What should be another PR? 
Buffering writer output until `\n`?

jnatalzia commented:
Re: Variadic method

For the first pass I feel like the slice approach is good enough but if it's not a huge ordeal we might as well put it in this PR. What do you think the LoE is?

jnatalzia commented on log.go:
```diff
@@ -0,0 +1,59 @@
+	for _, rule := range s.rules {
+		marked = Mark(marked, rule.Matcher, rule.Color)
+	}
+	return s.out.Write([]byte(marked))
```
thoughts on a convenience `Write` method that takes in a string?

cyucelen replied:
What do you mean by that? 
```golang
type Writer interface {
	Write(p []byte) (n int, err error)
}
```
The `io.Write` takes `[]byte`, if we want to be a `Writer` we need to play the game by the rules :smile:

jnatalzia replied:
We can extend that interface right? I'm still learning some of the golang internals but my understanding was we had to match the interface but could add additional functionality on top of it. So I was suggesting a `WriteString` func which takes a string, transforms it into a byte array then calls `Write` but if this is bad practice I'm happy to skip!

cyucelen replied:
Why should we do that? Most probably user will not deal with the `Write` function, it is just for integrating with `log.Logger`. He/she just can create `StdoutMarker` instance, add the rules and pass it to the `log.Logger` of Golang as Output. (_log.Logger accepts output stream type of **io.Writer**_). 
Then Golang's logger output will be redirected to us and we will mark it before sending it to the os.Stdout!

Simply:
```golang
stdoutMarker := marker.NewStdoutMarker()
stdoutMarker.AddRules(...)
logger := log.New(stdoutMarker, "", 0) // We can pass stdoutMarker as output, since we implemented io.Writer

logger.Println(...) // All
logger.Print(...) // These
logger.Sprintf(...) // are Marked!
// no need to use stdoutMarker.Write(..)
```   
 :tada: 

Since user will not deal with `Write` stuff, i believe there is no need for `WriteString`(unless we are integrating with something that accepts such interface) to prettify our API. Correct me if I get it wrong.

jnatalzia replied:
Makes sense!

jnatalzia commented on log_test.go:
```diff
@@ -0,0 +1,71 @@
+	stdoutMarker := NewStdoutMarker()
+	markRule := MarkRule{MatchAll("test"), color.New(color.FgRed)}
+	stdoutMarker.AddRule(markRule)
+	assert.Len(t, stdoutMarker.rules, 1)
```
is it overkill to test the object in the rules is what we expect it to be?

cyucelen replied:
assertion fails when i do that.
```
assert.Equal(t, stdoutMarker.rules[0], markRule)
```
Output:
```
expected: marker.MarkRule{Matcher:(marker.MatcherFunc)(0x6a1780), Color:(*color.Color)(0xc0000af200)}
actual  : marker.MarkRule{Matcher:(marker.MatcherFunc)(0x6a1780), Color:(*color.Color)(0xc0000af200)}
```
Most probably because of, in `testify/assert` there is a statement for `assert.Equal`: 
> Function equality cannot be determined and will always fail. 

For ensuring:
```
assert.Equal(t, markRule.Matcher, stdoutMarker.rules[0].Matcher)
```
Output:
```
Error:      	Invalid operation: (marker.MatcherFunc)(0x6a1710) == (marker.MatcherFunc)(0x6a1710) (cannot take func type as argument)
```

cyucelen replied:
Just found a way to assert function equality with a little spice of `reflect`. :bowtie:

Darkclainer replied:
You can not compare function, nor you can take their address - it is prohibited in standard. You can wrap it in interface, I think.

jnatalzia commented on log_test.go:
```diff
@@ -0,0 +1,71 @@
+		{MatchEmail(), color.New(color.FgCyan)},
+	}
+	stdoutMarker.AddRules(markRules)
+	assert.Len(t, stdoutMarker.rules, 3)
```
same q as L35

cyucelen commented:
Oh, we need to test the rule apply order by the way. Since we are applying the rules sequentially, some matches can overlap - later order matcher will override the previous ones if there is an overlap.

Users need to be aware of it and add their mark rules according to the order too. We need to document this and give an example for it too.

cyucelen commented:
> Re: Variadic method
> 
> For the first pass I feel like the slice approach is good enough but if it's not a huge ordeal we might as well put it in this PR. What do you think the LoE is?

I don't know what `LoE` is :smile: but we can skip it for now, less is more!

cyucelen commented:
Implemented some helper functionality for our assertions in `assert.go` and tested mark rules with it. Can you check it out? Then we can resolve the corresponding reviews.

Darkclainer commented:
You can not compare function. See https://stackoverflow.com/questions/42140758/collection-of-unique-functions-in-go/42147285#42147285

cyucelen commented:
> You can not compare function. See https://stackoverflow.com/questions/42140758/collection-of-unique-functions-in-go/42147285#42147285

[This](https://stackoverflow.com/a/45134241) approach actually looks like working. I tested it for our `MatcherFunc` type and it fails when `MatcherFunc`s actually not equal.

Darkclainer commented:
This behaviour is implementation defined.

cyucelen commented:
> This behaviour is implementation defined.

You mean **undefined**?

Darkclainer commented:
No. Current standard says that functions are incomparable, see https://golang.org/ref/spec#Comparison_operators . 
Using reflect in that particular case, with current go implementation you got desired behavior.

cyucelen commented:
I get it.

This behavior is not specified and so, can vary between different compiler implementations.
Here is the question; is it really a big problem for us? 

- If it is, then, is there any other option for directly testing the `AddRules` behaviour via asserting `MarkRule` equality. (You said something about wrapping it with interface?)
- Or, we can test it indirectly in `Write` functions' tests, this is fine too. (I mean just testing the behavior of `Write` can cover `AddRules` logic too)

I prefer the second way if the tests for the first option will impact the implementation and make it more complex.

Darkclainer commented:
I think second option is fine, but you should come up with test that cover order of rules too.

cyucelen commented:
Aye aye, captain!

jnatalzia commented on log_test.go:
```diff
@@ -0,0 +1,61 @@
+
+	expectedLog = fmt.Sprintf("best %s %s is %s\n", red("data"), red("company"), red(blue("skydome")))
+	logger.Print("best data company is skydome")
+	assert.Equal(t, expectedLog, mockOut.actualLog)
```
nice, this is so cool

jnatalzia commented:
@cyucelen is it just the readme docs that remain for this feature? super excited to integrate it into some projects

cyucelen commented:
@jnatalzia yes, can you help about docs?

jnatalzia commented:
Can do!

jnatalzia commented on log.go:
```diff
@@ -0,0 +1,57 @@
+}
+
+// NewWriteMarker creates a Marker that writes out to the given io.Writer
+func NewWriteMarker(writer io.Writer, options ...StdoutMarkerOption) *StdoutMarker {
```
hey @cyucelen in writing the docs, I refactored the way that we deal with custom `out`s. instead of having to replace stdout on the created marker, i flipped it so both cases (stdout and custom) require a single function call and no more than that.

let me know what you think, happy to discuss changing it back but it felt more natural when making the example file for this writer type so i thought it might be a good change

jnatalzia commented:
Docs added! Also refactored a small portion of the code, added a comment explaining what and why to the PR

cyucelen commented:
We merged the @Darkclainer s unified image PR to master. Can you integrate it to this branch? I will review your last commit about Writer asap!

jnatalzia commented:
Yes I can! Also seems like I need to add some code coverage, gotta love CI. Will fix it up today at some point

jnatalzia commented:
Hmmm @Darkclainer  can we make an issue/PR to update the docs for using the image generator? It requires a pip install (maybe foreign for a go repo) as well as two separate terminal dependencies but no indication of where to get them. I can guess but given I'm installing them in the shell, I'd prefer to have sanctioned sources

cc @cyucelen

cyucelen commented:
Awesome PR @jnatalzia! 

I just added some docs for image_generator, also updated the images of log feature in readme. 

Maybe we can get rid of functional options in `WriteMarker` since we are not using it anymore ?

Looks like we are getting ready for the deployment! :sunglasses: 

Also, what do you think about #29 ? @jnatalzia  @Darkclainer

jnatalzia commented:
> Maybe we can get rid of functional options in WriteMarker since we are not using it anymore ?

Sounds like a plan! Will push an update with those removed and then I think we're good to merge. Will mark as no longer draft.

Will comment on #29 on the issue!

cyucelen commented:
Looks like we are good to go! Do the final touch, merge it!
